### DISCUSSION TITLE
Integration with git-dependent tools

### DISCUSSION THREAD

**Discussion by kriomant** [jj-vcs/jj#4693] [General]
I started using jj and I love it. However, I miss integration with tools which expects git repository.
For example, I really miss `cargo release`. It dislikes detached checkouts.

Is it possible to create wrapper for such tools? How I see it:

You run `jj git run cargo release`, jj assigns git branch to current head, starts given tool, then absorbs changes made by tool.

**Comment by PhilipMetzger** [comment]
> I started using jj and I love it. However, I miss integration with tools which expects git repository.
> For example, I really miss `cargo release`. It dislikes detached checkouts.

Then you should colocate by default. Cargo at the moment won't accept any new VCS's, see https://internals.rust-lang.org/t/support-for-jj-jujutsu-vcs/19250

**Reply by martinvonz** [reply]
To clarify, "colocate" refers to https://martinvonz.github.io/jj/latest/git-compatibility/#co-located-jujutsugit-repos

**Reply by kriomant** [reply]
I use colocation, it doesn't help in this case.

**Comment by tim-janik** [comment]
> I started using jj and I love it. However, I miss integration with tools which expects git repository. For example, I really miss `cargo release`. It dislikes detached checkouts.
> 
> Is it possible to create wrapper for such tools? How I see it:
> 
> You run `jj git run cargo release`, jj assigns git branch to current head, starts given tool, then absorbs changes made by tool.

First, colocation doesn't fix Git tools not expecting a detached head state, for proper interop you also need to have JJ support for properly advancing Git branches, as outlined here:   [Branch Interoperability](https://github.com/martinvonz/jj/pull/4670#issuecomment-2428118457)

In the same PR, @martinvonz asked to collect use cases for better interoperability / Integration with git-dependent tools, it might sense to list your case there: [scenarios that should improve](https://github.com/martinvonz/jj/pull/4670#discussion_r1807534569)

**Reply by kriomant** [reply]
This is useful link, thank you!

**Comment by ilyagr** [comment]
In a co-located repo, if the detached head is a problem for `cargo release`, you could switch to a branch with git (e.g. `git switch branch`) before running `cargo release`.

`jj` should handle that just fine; if it doesn't, please let us know about any problems.

As others said, there are also discussion of having `jj` be able to preserve the current branch somehow, but that is not currently implemented, so you might have to redo `git switch` after mutating `jj` commands.

**Reply by kriomant** [reply]
Thank you, I'll try this.
